File: lecture06/stack_sorting.py
#!/usr/bin/env python3

# Functions

def sort_stack(old_stack):
    new_stack = []

    while old_stack:
        t = old_stack.pop(-1)

        # Go through new stack and copy back over values greater than temporary
        while new_stack and old_stack[-1] > t:
        # while new_stack and new_stack[-1] > t:    # Fix
            old_stack.append(new_stack.pop(-1))

        # Place temporary value on top
        new_stack.append(t)

    return new_stack

# Main execution

def main():
    stack = [5, 4, 7, 0, 1]
    print(stack)
    print(sort_stack(stack))

if __name__ == '__main__':
    main()

File: lecture06/stack_sorting.cpp
// stack_sorting.cpp

#include <iostream>
#include <stack>

using namespace std;

stack<int> sort_stack(stack<int> &os) {
    stack<int> ns;

    while (!os.empty()) {
    	auto t = os.top(); os.pop();
    	while (!ns.empty() && os.top() > t) {
    	// while (!ns.empty() && ns.top() > t) { // FIX
    	    os.push(ns.top());
    	    ns.pop();
	}

	ns.push(t);
    }

    return ns;
}

int main(int argc, char *argv[]) {
    stack<int> s;

    s.push(5);
    s.push(4);
    s.push(7);
    s.push(0);
    s.push(1);

    s = sort_stack(s);
    while (!s.empty()) {
    	cout << s.top() << endl;
    	s.pop();
    }

    return EXIT_SUCCESS;
}
